doc/BGP/Bgpcfgd-dyn-peer-modification-support.md

```diff
@@ -0,0 +1,144 @@
+
+## 1.2 Scalability requirements
+###### Table 2: Scalability
+| Component                | Expected value              |
+|--------------------------|-----------------------------|
+| Number of Dynamic BGP Peers each with route-map, prefix-list and peer-group|2k|
+| Number of listen ranges|4k|
+| Size of each Route-map, Prefix-list|<10|
+
+
+# 2 Modules Design
+
+## 2.1 Config DB
+Existing Config DB tables will be used to achieve this, no new Config DB table is introduced
+```
+BGP_PEER_RANGE|{{VRF/VNET-name}}|{{Peer-name}}:
+    "ip_range": [{{Range of IPs to add to listen range}}],
+    "name": {{Peer-name}},
+    "peer_asn": {{Peer's ASN number}}, (Optional)
+    "src_address": {{Src IP to initiate session}} (Optional)
+```
+
+## 2.2 State DB
+Following new table will be added to State DB.
+Use case: The SDN controller which programs neighbors will query the following state DB entry to confirm if the required entry is processed by bgpcfgd and programmed successfully. 
+```
+BGP_PEER_CONFIGURED_TABLE|{{VRF/VNET-name}}|{{Peer-name}}:
+    "ip_range": [{{Range of IPs to add to listen range}}],
+    "name": {{Peer-name}},
+    "peer_asn": {{Peer's ASN number}}, (Optional)
+    "src_address": {{Src IP to initiate session}} (Optional)
+```
+Unless otherwise stated, the attributes are mandatory.
+
+While the above shows the State DB schema for a dynamic peer, the schema for static peers bear the same table name format of ```BGP_PEER_CONFIGURED_TABLE|{{VRF/VNET-name}}|{{Peer-name}}```, but the key value pairs under the table will be those which are used to configure static peers.
+
+## 2.3 CLI
+The following CLIs will be added, each of these will have similar CLI outputs to their ```default VRF``` counterparts which are widely in use today:
+```
+1. show ip bgp vrf {vrf_name} summary
+2. show ip bgp vrf {vrf_name} network
+3. show ip bgp vrf {vrf_name} neighbors
+4. show ipv6 bgp vrf {vrf_name} summary
+5. show ipv6 bgp vrf {vrf_name} network
+6. show ipv6 bgp vrf {vrf_name} neighbors
+```
+
+## 2.4 Bgpcfgd
+The following modifications will be made to bgpcfgd to support the new scenario:
+1. BGPPeerMgrBase will check, at init time, if an update template file exists as follows ```"bgpd/templates/" + self.constants["bgp"]["peers"][peer_type]["template_dir"] + "/" + update.conf.j2```.  
+   The `update.conf.j2` file is a Jinja2 template used to define the configuration updates for dynamic BGP peers.
+2. If one exists, then we will consider this device as supporting updates to the peer_type in question, and we append the update template to the list of templates which can be rendered:
+```
+        self.templates["update"] = self.fabric.from_file(update_template_path)
+```
+3. If an update template is supported and the peer_type is dynamic, we will identify the set of ip_ranges which are added, and those which are deleted as part of the operation, and pass in those as additional kwargs and render the update template, sample code:
```

Comment:
> we will identify the set of ip_ranges which are added, and those which are deleted as part of the operation

Is this done in bgpcfgd? how is this done? Can you add more details?

Reply:
Yes it is done in bgpcfgd, it is done by querying vtysh to find the current set of ranges, and diff-ing that against the newly configured set of ranges, the added/deleted ranges are then passed into an update.j2 template to render the required config changes. PR: https://github.com/sonic-net/sonic-buildimage/pull/22261

I will update the HLD PR to add some more details on how it is done in bgpcfgd